File: worker/src/main/java/com/github/k0kubun/github_ranking/repository/PaginatedUsers.java
package com.github.k0kubun.github_ranking.repository;

import com.github.k0kubun.github_ranking.repository.dao.UserDao;
import com.github.k0kubun.github_ranking.model.User;

import java.util.List;

import org.skife.jdbi.v2.Handle;

// This class does cursor-based-pagination for users order by stargazers_count DESC.
public class PaginatedUsers
{
    private static final int PAGE_SIZE = 5000;

    private final UserDao userDao;
    private Integer lastMinStars;
    private Integer lastMinId;

    public PaginatedUsers(Handle handle)
    {
        userDao = handle.attach(UserDao.class);
        lastMinStars = null;
        lastMinId = null;
    }

    public List<User> nextUsers()
    {
        List<User> users;
        if (lastMinId == null && lastMinStars == null) {
            users = userDao.starsDescFirstUsers(PAGE_SIZE);
        }
        else {
            users = userDao.starsDescUsersAfter(lastMinStars, lastMinId, PAGE_SIZE);
        }
        if (users.isEmpty()) {
            return users;
        }

        User lastUser = users.get(users.size() - 1);
        lastMinStars = lastUser.getStargazersCount();
        lastMinId = lastUser.getId();
        return users;
    }
}

File: worker/src/main/java/com/github/k0kubun/gitstar_ranking/repository/PaginatedUsers.kt
package com.github.k0kubun.gitstar_ranking.repository

import com.github.k0kubun.gitstar_ranking.model.User
import com.github.k0kubun.gitstar_ranking.repository.dao.UserDao
import org.skife.jdbi.v2.Handle

// This class does cursor-based-pagination for users order by stargazers_count DESC.
class PaginatedUsers(handle: Handle) {
    private val userDao: UserDao = handle.attach(UserDao::class.java)
    private var lastMinStars: Int?
    private var lastMinId: Long?
    fun nextUsers(): List<User> {
        val users: List<User> = if (lastMinId == null && lastMinStars == null) {
            userDao.starsDescFirstUsers(PAGE_SIZE)
        } else {
            userDao.starsDescUsersAfter(lastMinStars, lastMinId, PAGE_SIZE)
        }
        if (users.isEmpty()) {
            return users
        }
        val lastUser = users[users.size - 1]
        lastMinStars = lastUser.stargazersCount
        lastMinId = lastUser.id
        return users
    }

    companion object {
        private const val PAGE_SIZE = 5000
    }

    init {
        lastMinStars = null
        lastMinId = null
    }
}
